samples/nrf9160/modem_shell/src/drivers/nrf91_non_offloaded_dev.c

```diff
@@ -0,0 +1,495 @@
+#define AT_CMD_PDP_CONTEXT_READ_DYN_INFO_DNS_ADDR_SECONDARY_INDEX 7
+#define AT_CMD_PDP_CONTEXT_READ_DYN_INFO_MTU_INDEX 12
+
+static int util_get_pdn_conn_dyn_params(int cid, struct pdn_conn_dyn_params *ret_dyn_info)
+{
+	struct at_param_list param_list = { 0 };
+	size_t param_str_len;
+	char *next_param_str;
+	char at_cmd_str[16];
+	char at_cmd_response_str[512];
+	char dns_addr_str[128];
+	char *at_ptr = at_cmd_response_str;
+	int ret, family, iterator;
+	struct in_addr *addr;
+	struct in6_addr *addr6;
+	bool resp_continues = true;
+
+	sprintf(at_cmd_str, AT_CMD_PDP_CONTEXT_READ_DYN_INFO, cid);
+
+	ret = nrf_modem_at_cmd(at_cmd_response_str, sizeof(at_cmd_response_str), at_cmd_str);
+	if (ret) {
+		mosh_error("Cannot get PDP conn dyn params, ret: %d", ret);
+		return false;
+	}
+
+	/* Parse the response */
+	ret = at_params_list_init(&param_list, AT_CMD_PDP_CONTEXT_READ_DYN_INFO_PARAM_COUNT);
+	if (ret) {
+		mosh_error("Could not init AT params list, error: %d\n", ret);
+		return ret;
+	}
+
+	iterator = 0;
+	while (resp_continues) {
+		resp_continues = false;
+		ret = at_parser_max_params_from_str(at_ptr, &next_param_str, &param_list,
+						    AT_CMD_PDP_CONTEXT_READ_DYN_INFO_PARAM_COUNT);
+		if (ret == -EAGAIN) {
+			resp_continues = true;
+		} else if (ret == -E2BIG) {
+			mosh_error("E2BIG, error: %d\n", ret);
+		} else if (ret != 0) {
+			mosh_error("Could not parse AT response for %s, error: %d\n", at_cmd_str,
+				   ret);
+			goto clean_exit;
+		}
+
+		/* Read primary DNS address */
+		param_str_len = sizeof(dns_addr_str);
+		ret = at_params_string_get(&param_list,
+					   AT_CMD_PDP_CONTEXT_READ_DYN_INFO_DNS_ADDR_PRIMARY_INDEX,
+					   dns_addr_str, &param_str_len);
+		if (ret) {
+			mosh_error("Could not parse dns str for cid %d, err: %d", cid, ret);
+			goto clean_exit;
+		}
+		dns_addr_str[param_str_len] = '\0';
+
+		family = net_utils_sa_family_from_ip_string(dns_addr_str);
+		if (family == AF_INET) {
+			addr = &(ret_dyn_info->dns_addr4_primary);
+			(void)inet_pton(AF_INET, dns_addr_str, addr);
+		} else if (family == AF_INET6) {
+			addr6 = &(ret_dyn_info->dns_addr6_primary);
+			(void)inet_pton(AF_INET6, dns_addr_str, addr6);
+		}
+
+		/* Read secondary DNS address */
+		param_str_len = sizeof(dns_addr_str);
+
+		ret = at_params_string_get(
+			&param_list, AT_CMD_PDP_CONTEXT_READ_DYN_INFO_DNS_ADDR_SECONDARY_INDEX,
+			dns_addr_str, &param_str_len);
+		if (ret) {
+			mosh_error("Could not parse dns str, err: %d", ret);
+			goto clean_exit;
+		}
+		dns_addr_str[param_str_len] = '\0';
+
+		family = net_utils_sa_family_from_ip_string(dns_addr_str);
+		if (family == AF_INET) {
+			addr = &(ret_dyn_info->dns_addr4_secondary);
+			(void)inet_pton(AF_INET, dns_addr_str, addr);
+		} else if (family == AF_INET6) {
+			addr6 = &(ret_dyn_info->dns_addr6_secondary);
+			(void)inet_pton(AF_INET6, dns_addr_str, addr6);
+		}
+
+		/* Read link MTU if exists: */
+		if (iterator == 1) {
```

Comment:
With IPv4v6 there's two `+CGCONTRDP` lines, first contains MTU for IPv4 and second for IPv6, I think. But does this logic work with IPv6 only? Wouldn't there be only a single `+CGCONTRDP` line in that case and it would contain IPv6 MTU?

Reply:
Good point. Possible if returning like that in that case. However, couldn't test that case. In DNA, for plain ipv6 PDN, it is always returning 2 lines because there are always also ipv4 DNS servers.
IPv6 is still TODO for this non-offloading device.

Reply:
It seems that the AT command spec is a bit vague for this command. It says that the MTU parameter is "IPv4 MTU", but my assumption is that if the DNS addresses are for IPv6, also the MTU is for IPv6. If that's the case, then I think the only way to detect whether the line contains IPv4 or IPv6 MTU would be to check the DNS addresses. Just wondering whether the DNS addresses are mandatory... If not, then it gets even more complicated.

Reply:
OK, changed the logic so that primary DNS address is used to detect the family for MTU. If no primary DNS address, then default MTU is used (1500).